packages/devtools_app/lib/src/timeline/timeline_flame_chart.dart

```diff
@@ -333,6 +413,119 @@ class FullTimelineFlameChartCanvas extends FlameChartCanvas<FullTimelineData> {
         0;
     return absoluteY - topOffset - (section * sectionSpacing);
   }
+
+  @override
+  void paintCallback(CanvasRenderingContext2D canvas, Rect visible) {
+    paintSections(canvas, visible);
+    paintRows(canvas, visible);
+    _paintAsyncGuidelines(canvas, visible);
+    paintTimelineGrid(canvas, visible);
+  }
+
+  void _paintAsyncGuidelines(CanvasRenderingContext2D canvas, Rect visible) {
+    final firstVerticalGuideline = _binarySearchForVerticalLineSegment(visible);
+    final firstHorizontalGuideline =
+        _binarySearchForHorizontalLineSegment(visible);
+
+    // Only modify the canvas style if we have any guidelines to paint.
+    if (firstHorizontalGuideline != null || firstVerticalGuideline != null) {
+      canvas
+        ..strokeStyle = colorToCss(guidelineColor)
+        ..lineWidth = guidelineWidth;
+    }
+
+    if (firstVerticalGuideline != null) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        verticalGuidelines,
+        firstVerticalGuideline,
+      );
+    }
+
+    if (firstHorizontalGuideline != null) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        horizontalGuidelines,
+        firstHorizontalGuideline,
+      );
+    }
+  }
+
+  void _paintGuidelines(
+    CanvasRenderingContext2D canvas,
+    Rect visible,
+    List<LineSegment> guidelines,
+    LineSegment first,
+  ) {
+    final guidelinesToPaint = guidelines
+        .sublist(guidelines.indexOf(first))
+        .where((line) => line.intersects(visible));
```

Comment:
this is currently defeating the purpose of doing the binary search.
the indexOf check is itself O(n)
then the where close is also O(n) as it has to check every element in the list.
What you probably want instead is to 
pass in the index of the first match.
Then do a for loop through all the line indexes, breaking when you find the first guide line that doesn't match on the main axis and filtering out all guidelines that are out of range on the cross-axis.